scraping: limit detail on dropped targets, to save memory

bboreham opened the pull request:
Fixes #12482

It's possible (quite common on Kubernetes) to have a service discovery return thousands of targets then drop most of them in relabel rules. The main place this data is used is to display in the web UI, where you don't want thousands of lines of display.

Set a default limit of 100 per job to store full detail, which I picked to be consistent with #4212. It seems that behaviour, of limiting the number drawn in the UI, has disappeared.

Update 2: needed to extend the API to let the UI show correct counts.
Please note I know nothing at all about React programming.

juliusv commented on web/ui/react-app/src/pages/serviceDiscovery/Services.tsx:
```diff
@@ -10,10 +10,12 @@ import { API_PATH } from '../../constants/constants';
 import { KVSearch } from '@nexucis/kvsearch';
 import { Container } from 'reactstrap';
 import SearchBar from '../../components/SearchBar';
+import internal from 'stream';
```
Did this line get added by accident? Seems like it shouldn't be there.

bboreham replied:
Removed.

juliusv commented on config/config.go:
```diff
@@ -155,6 +155,7 @@ var (
 		HonorLabels:             false,
 		HonorTimestamps:         true,
 		HTTPClientConfig:        config.DefaultHTTPClientConfig,
+		KeepDroppedTargets:      100,
```
This reads like a boolean option, but I failed to quickly come up with a nice-sounding option that sounds more like a numeric limit, so I guess it's fine unless someone has a better idea :)

juliusv reviewed (commented):
Nice, I like it. But are we allowed to change the default? We publicly document this API at https://prometheus.io/docs/prometheus/latest/querying/api/#targets and we promise stability for the HTTP API at https://prometheus.io/docs/prometheus/latest/stability/#api-stability-guarantees.

juliusv commented:
I mean honestly I'm all for breaking minor things like this, but I also wouldn't be surprised if someone out there was using this API in some large scenario to get all discovered targets. @roidelapluie what do you think?

> Adding `droppedTargetCounts` is a change to the API, albeit backwards-compatible. Would that be allowed, or does this have to be on a new path?

That's totally fine, just breaking changes like returning fewer targets than before (by default) would technically be disallowed by our promises. Whether we want to follow those promises to that degree is a matter of debate...

bboreham commented:
I changed the default to 0, meaning no limit.
Also added as a global option, and added that to the Kubernetes example config.

Commits should be squashed before merging.

juliusv commented on docs/configuration/configuration.md:
```diff
@@ -106,6 +106,10 @@ global:
 
+  # Per-job limit on the number of targets dropped by relabeling
+  # that will be kept in memory. 0 means no limit.
+	[ keep_dropped_targets: <int> | default = 0 ]
```
There seems to be a tab indent here which should be a two-space indent.

juliusv commented on docs/configuration/configuration.md:
```diff
@@ -106,6 +106,10 @@ global:
   # change in the future.
   [ target_limit: <int> | default = 0 ]
 
+  # Per-job limit on the number of targets dropped by relabeling
```
This is not per job, but per scrape pool, which to the user is exposed as a scrape config (there can be multiple jobs per scrape config if targets have a `job` label set that overrides the scrape config's default job name). So I'd say "Per-scrape-config limit ..." or "Limit per scrape config ..." (less hyphenated) or something like that.

juliusv reviewed (commented):
:+1: besides last comments

roidelapluie reviewed (approved):
LGTM!

bboreham commented:
Note there was a bug, which required #12726.
